### DISCUSSION TITLE
Upload image/file and save it to filesystem no longer works same way, what is method to do this now?

### DISCUSSION THREAD

**Discussion by half-metal** [sveltejs/kit#7974] [Q&A] [answered]
I wrote about this issue here https://github.com/sveltejs/kit/issues/7924 where a previous upload method no longer works. @Conduitry  recommended writing here and Discord. Already created a topic in Discord. 

This repo https://github.com/Programonaut-Blog/sveltekit-image-upload is fine for updating an image in filesystem using a previous version of Sveltekit:

Something changed in Sveltekit, and a very common use case "write a file to filesystem" has what seems like no documentation for Sveltekit in its current form.  What's the recommended way to do this now? I would like the application to have the ability to upload an image and save on filesystem. This worked fine before, but all the examples online are using old Sveltekit and error out.

This explains using request.json() that I am used to https://stackoverflow.com/questions/70990443/unable-to-access-request-body-from-the-endpoint-js-in-sveltekit-skeleton-project, but I get an error with "Unexpected token o in JSON at position 1"

### Reproduction

Use this same code with latest version of Sveltekit, no longer works.
https://github.com/Programonaut-Blog/sveltekit-image-upload

### Logs
When attempting to use .json() with code above I get this
```Shell
Unexpected token o in JSON at position 1
SyntaxError: Unexpected token o in JSON at position 1
    at JSON.parse (<anonymous>)
```

### System Info

```Shell
Mac, svelte latest 1.0.0-next.571
```

### Severity

serious, but I can work around it

### Additional Information

Still boggled Sveltekit doesn't have core common use cases which would be updated with major changes, would help with adoption and probably cut down on defects.

**Comment by teemingc** [comment, marked as answer]
I've created an updated example you can find here: https://github.com/s3812497/svelte-kit-upload-image
The most notable change to get it working was migrating the old file-based routing to the new folder-based routing system.

Please see the following documents for any other migrations you may require:
* https://kit.svelte.dev/docs/routing
* https://github.com/sveltejs/kit/discussions/5774

**Reply by half-metal** [reply]
This works great with latest Sveltekit, thank you!

**Reply by ndru123** [reply]
@s3812497 this works, but not in production. If I use npm run build and then npm run preview the uploaded picture wont update. It will still upload the picture in the static directory though. Do you know hot to get it to work in production?

**Reply by teemingc** [reply]
Hi there @ndru123 , I've updated the example and it should work correctly now :)

**Reply by ndru123** [reply]
thank you, works great!

**Reply by ndru123** [reply]
@s3812497 do you have any idea how to get image upload to work when using a node Custom server (https://kit.svelte.dev/docs/adapter-node#custom-server) I'm trying to replicate your image-upload, but have to start with my node custom server and starting with "node server.js" instead of "npm run preview". My files get saved in an img folder in the build directory, but everytime new uploaded files are served it gives a 404.

**Reply by teemingc** [reply]
That depends on the directory the server is running in and the filepath the form action is saving the images to. You will need to identify these two (maybe by printing out the directory using `console.log` in your server/action?)

**Reply by ndru123** [reply]
@s3812497 yes, that's exactly what I did. When you start with "npm run preview", like in your project, files are served from and saved in .svelte-kit/output/client. When I run with "node server.js " files are served from and saved in build/client. I changed my app so that the images are saved in the build/client folder, which works fine. However, if I then want to serve the file I get a 404. Even when i put https://myappadress:port/ImageIJustUploaded.png it doesn't show up and gives me 404, even though the image is in the right directory. Really strange.

**Reply by Josh-H-555** [reply]
Did you ever find a way to serve files uploaded in npm run preview? Trying to keep things simple and just be able to upload an image into a folder, and then grab that image at runtime.

npm run dev works, but any production file uploads seems to require a rebuild.

**Reply by teemingc** [reply]
After a bit of hacking around, I've updated [the example](https://github.com/s3812497/svelte-kit-upload-image) to work using a custom node server. Just build and run `npm run start`.
The custom node server uses an [express middleware](https://github.com/lukeed/polka/tree/master/examples/with-serve-static) to serve the uploaded images. Previously, the server had to be restarted for the uploaded images to be recognised. This fixes that (uploaded images are served properly).

Hope this helps! Let me know how you guys go.

**Reply by clock1998** [reply]
@s3812497 I am surprised that the build cannot serve the files the way it should. I am using Nginx, so I just use it to serve my files.

**Reply by teemingc** [reply]
> @s3812497 I am using Nginx, so I just use it to serve my files. 

I think this is definitely the ideal way of serving the files.

**Comment by ayuancool** [comment]
Why can't we provide an out of the box 'upload' directory?

**Reply by david-plugge** [reply]
Because you might want to upload files to another service like s3 instead of saving it to the local filesystem. There are too many options to consider and therefore it is left to the user. Same for things like database, authentication and so on. Its not sveltekits responsibility to do these things

**Reply by ayuancool** [reply]
The significance of database, authentication, and other tasks is different from that of this temporary file transfer station. The former can be entrusted to a third party for processing, but the latter is likely a meaningless public access resource, just a temporary folder for us to easily manipulate file preprocessing work such as decompression and content review. In short, the 'static' folder during development disappears after 'build', losing the ability to quickly add and reference some content. Of course, as you said, it is not necessary for sveltekit, and there are many solutions
